libs/runtime/ebpf_epoch.c

```diff
@@ -4,130 +4,171 @@
+    ebpf_list_entry_t free_list;           ///< Per-CPU free list.
+    int64_t current_epoch;                 ///< The current epoch for this CPU.
+    int64_t released_epoch;                ///< The newest epoch that can be released.
+    int timer_armed : 1;                   ///< Set if the flush timer is armed.
+    int rundown_in_progress : 1;           ///< Set if rundown is in progress.
+    int epoch_computation_in_progress : 1; ///< Set if epoch computation is in progress.
 } ebpf_epoch_cpu_entry_t;
 
-C_ASSERT(sizeof(ebpf_epoch_cpu_entry_t) % EBPF_CACHE_LINE_SIZE == 0);            // Verify alignment.
-C_ASSERT(EBPF_EPOCH_THREAD_TABLE_SIZE > EBPF_EPOCH_THREAD_TABLE_RESERVED_COUNT); // Verify that the reserved count fits.
-
+/**
+ * @brief Table of per-CPU state.
+ */
 static _Writable_elements_(_ebpf_epoch_cpu_count) ebpf_epoch_cpu_entry_t* _ebpf_epoch_cpu_table = NULL;
+
+/**
+ * @brief Number of CPUs in the system as determined at initialization time.
+ */
 static uint32_t _ebpf_epoch_cpu_count = 0;
 
 /**
- * @brief _ebpf_current_epoch indicates the newest active epoch. All memory free
- * operations were performed prior to this value.
+ * @brief Enum of messages sent between CPUs.
+ */
+typedef enum _ebpf_epoch_cpu_message_type
+{
+    EBPF_EPOCH_CPU_MESSAGE_TYPE_PROPOSE_RELEASE_EPOCH, ///< This message is sent to CPU 0 to propose a new release
+                                                       ///< epoch.
+                                                       ///< CPU 0 declares the new current epoch and proposes it as the
+                                                       ///< release epoch. Each CPU then queries the epoch for each
+                                                       ///< thread linked to this CPU and sets the proposed release
+                                                       ///< epoch in the message to the minimum of the local minima and
+                                                       ///< the minima in the message. The message is then forwarded to
+                                                       ///< the next CPU. The last CPU then sends a epoch commit message
+                                                       ///< to CPU 0 with the final proposed release epoch.
+
+    EBPF_EPOCH_CPU_MESSAGE_TYPE_COMMIT_RELEASE_EPOCH, ///< This message is sent to CPU 0 to commit the proposed release
+                                                      ///< epoch.
+                                                      ///< Each CPU then:
+                                                      ///< 1. Clears the timer-armed flag.
+    ///< 2. Sets the released epoch to the proposed release epoch minus 1.
+    ///< 3. Releases any items in the free list that are eligible for reclamation.
+    ///< 4. Rearms the timer if need.
+    ///< 5. Forwards the message to the next CPU.
+    ///< The last CPU then sends a epoch computation complete message to CPU 0.
```

Comment:
Seems odd that these don't line up with line 80

Reply:
Clang-format overrides it and moves them back. I can override the formatting, but clang-format will revert it.